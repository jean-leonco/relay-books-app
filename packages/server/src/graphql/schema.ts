import { GraphQLSchema } from 'graphql';

import applyMiddlewares from '../core/middlewares/applyMiddlewares';

// import SubscriptionType from './subscription/SubscriptionType';
import MutationType from './type/MutationType';
import QueryType from './type/QueryType';

const _schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
  // subscription: SubscriptionType,
});

applyMiddlewares(_schema);

export const schema = _schema;
