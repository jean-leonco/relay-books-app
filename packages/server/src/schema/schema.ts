import { GraphQLSchema } from 'graphql';

import applyMiddlewares from '../graphql/middlewares/applyMiddlewares';

import MutationType from './MutationType';
import QueryType from './QueryType';

const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
});

applyMiddlewares(schema);

export default schema;
