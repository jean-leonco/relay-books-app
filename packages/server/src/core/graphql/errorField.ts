import { GraphQLString } from 'graphql';

const errorField = {
  error: {
    type: GraphQLString,
    description: 'Default error field resolver for mutations',
    resolve: ({ error }) => error,
  },
};

export default errorField;
