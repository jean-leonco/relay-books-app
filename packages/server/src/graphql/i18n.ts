import i18next from 'i18next';
import { Middleware } from 'koa';

import englishTranslation from '../locales/en';
import portugueseTranslation from '../locales/pt';

// Initialize i18n
i18next.init({
  resources: {
    en: englishTranslation,
    pt: portugueseTranslation,
  },
  preload: ['en', 'pt'],
  fallbackLng: 'en',
  nsSeparator: '::',
  whitelist: ['en', 'pt'],
  nonExplicitWhitelist: true,
});

export const i18nMiddleware: Middleware = async (context, next) => {
  const { lang } = context.header;

  if (lang) await i18next.changeLanguage(lang);

  return next();
};

export default i18next;
