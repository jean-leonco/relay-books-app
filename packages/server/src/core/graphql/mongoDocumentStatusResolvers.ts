import { GraphQLBoolean, GraphQLString } from 'graphql';

export const mongoDocumentStatusResolvers = {
  isActive: {
    type: GraphQLBoolean,
    description: 'The soft delete status.',
    resolve: (obj) => obj.isActive,
  },
  removedAt: {
    type: GraphQLString,
    description: 'The hard delete status.',
    resolve: (obj) => (obj.removedAt ? obj.removedAt.toISOString() : null),
  },
  createdAt: {
    type: GraphQLString,
    description: 'The date that the node was created.',
    resolve: (obj) => (obj.createdAt ? obj.createdAt.toISOString() : null),
  },
  updatedAt: {
    type: GraphQLString,
    description: 'The date that the node was last updated.',
    resolve: (obj) => (obj.updatedAt ? obj.updatedAt.toISOString() : null),
  },
};
