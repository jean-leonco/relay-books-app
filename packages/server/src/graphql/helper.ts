import { DocumentNode } from 'graphql/language';
import DataLoader from 'dataloader';

interface LoaderFactory<T extends DataLoader<any, any>> {
  getLoader: () => T;
}

interface LoaderFactoryMap {
  [key: string]: LoaderFactory<any>;
}

export type ResolvedLoaders<T extends LoaderFactoryMap> = { [K in keyof T]: ReturnType<T[K]['getLoader']> };

export function getDataloaders<T extends LoaderFactoryMap>(loaders: T): ResolvedLoaders<T> {
  const result: ResolvedLoaders<T> = {} as any;
  for (const key in loaders) {
    if (loaders[key].getLoader) {
      result[key] = loaders[key].getLoader();
    }
  }
  return result;
}

export function hasIntrospectionQuery(document: DocumentNode) {
  if (document.definitions.length === 0) {
    return false;
  }
  const definition = document.definitions[0];
  if (definition.kind !== 'OperationDefinition') {
    return false;
  }
  return definition.name && definition.name.value === 'IntrospectionQuery';
}
