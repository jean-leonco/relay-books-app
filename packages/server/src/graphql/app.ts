/* eslint-disable no-console */
import 'isomorphic-fetch';

import { koaPlayground } from 'graphql-playground-middleware';
import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import convert from 'koa-convert';
import cors from '@koa/cors';
import graphqlHttp from 'koa-graphql';
import koaLogger from 'koa-logger';
import Router from '@koa/router';
import idx from 'idx';
import i18next from 'i18next';
import koaI18next from 'koa-i18next';
import NoIntrospection from 'graphql-disable-introspection';
import { GraphQLError } from 'graphql';

import * as graphqlLoaders from '../loader';

import { JWT_KEY } from '../common/config';

import { KoaContextExt } from '../types';

import UserModel, { IUser } from '../modules/user/UserModel';

import { t } from '../locales/helpers';

import appPlatformMiddleware from '../modules/auth/appPlatformMiddleware';
import authMiddleware from '../modules/auth/authMiddleware';

import { getDataloaders } from './helper';
import { schema } from './schema';
import { i18nMiddleware } from './i18n';

const app = new Koa<any, KoaContextExt>();

if (process.env.NODE_ENV === 'production') {
  app.proxy = true;
}
app.keys = [JWT_KEY];

const loaders = graphqlLoaders;

const router = new Router<any, KoaContextExt>();

app.use(bodyParser());

app.on('error', (error) => {
  console.error('Error while answering request', { error });
});

if (process.env.NODE_ENV !== 'test') {
  app.use(koaLogger());
}

app.use(convert(cors({ maxAge: 86400, origin: '*' })));

router.get('/healthz', async (ctx) => {
  try {
    await UserModel.find().lean<IUser>();
    ctx.body = 'Database working';
    ctx.status = 200;
  } catch (err) {
    ctx.body = err.toString();
    ctx.status = 500;
  }
});

if (process.env.NODE_ENV !== 'production') {
  router.all(
    '/playground',
    koaPlayground({
      endpoint: '/graphql',
    }),
  );
}

// Middleware to get dataloaders
app.use(async (ctx, next) => {
  ctx.dataloaders = getDataloaders(loaders);
  await next();
});

app.use(
  koaI18next(i18next, {
    lookupCookie: 'lang', // detecting language in cookie
    order: ['header', 'cookie'],
    next: true,
  }),
);

app.use(i18nMiddleware);

// avoid requests from unknown clients
if (process.env.NODE_ENV !== 'development') {
  app.use(appPlatformMiddleware);
}

// Middleware to get result from authorization token
app.use(authMiddleware);

const graphqlSettingsPerReq = async (request, ctx, koaContext) => {
  const { dataloaders, user } = koaContext;
  const { appplatform, timezone } = request.header;

  if (process.env.NODE_ENV !== 'test') {
    console.info('Handling request', {
      appplatform,
      unauthorized: !user,
      userId: idx(user, (_) => _._id),
    });
  }

  return {
    graphiql: process.env.NODE_ENV !== 'production',
    schema,
    //rootValue: { request: ctx.req },
    validationRules: process.env.NODE_ENV === 'production' ? [NoIntrospection] : [],
    context: {
      user,
      dataloaders,
      appplatform,
      koaContext,
      timezone,
      t,
    },
    formatError: (error: GraphQLError) => {
      if (error.name && error.name === 'BadRequestError') {
        ctx.status = 400;
        ctx.body = 'Bad Request';
        return {
          message: 'Bad Request',
        };
      }

      if (error.path || error.name !== 'GraphQLError') {
        console.error(error);
      } else {
        console.log(`GraphQLWrongQuery: ${error.message}`);
      }

      console.error('GraphQL Error', { error });

      if (process.env.NODE_ENV !== 'production') {
        return {
          message: error.message,
          locations: error.locations,
          stack: error.stack,
        };
      } else {
        ctx.status = 400;
        ctx.body = 'Bad Request';
        return {
          message: 'Bad Request',
        };
      }
    },
  };
};

const graphqlServer = graphqlHttp(graphqlSettingsPerReq);
router.all('/graphql', graphqlServer);

app.use(router.routes()).use(router.allowedMethods());

export default app;
