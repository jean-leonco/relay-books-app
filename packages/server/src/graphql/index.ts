/* eslint-disable no-console */
import 'core-js/stable';
import { createServer } from 'http';

import { execute, subscribe } from 'graphql';
import { SubscriptionServer } from 'subscriptions-transport-ws';

import { GRAPHQL_PORT, GRAPHQL_HOST } from '../common/config';
import connectDatabase from '../common/database';

import app from './app';
import { schema } from './schema';

type ConnectionParams = {
  authorization?: string;
};

const runServer = async () => {
  try {
    console.log('\n🔗 Connecting to database...');
    await connectDatabase();
  } catch (error) {
    console.error('Could not connect to database', { error });
    throw error;
  }

  const server = createServer(app.callback());

  server.listen(GRAPHQL_PORT, () => {
    console.info(`\n🚀 Server started at http://${GRAPHQL_HOST}:${GRAPHQL_PORT}`);

    if (process.env.NODE_ENV !== 'production') {
      console.info(`\n🎠 GraphQL Playground available at http://${GRAPHQL_HOST}:${GRAPHQL_PORT}/playground\n`);
    }

    SubscriptionServer.create(
      {
        onConnect: async (connectionParams: ConnectionParams) => {
          console.info('Client subscription connected', connectionParams);
          // @TODO - get user from connectionParams?.authorization
          // @TODO - return context with user
        },
        onDisconnect: () => console.info('Client subscription disconnected'),
        execute,
        subscribe,
        schema,
      },
      {
        server,
        path: '/subscriptions',
      },
    );
  });
};

(async () => {
  console.log('\n📡 Server starting...');
  await runServer();
})();
