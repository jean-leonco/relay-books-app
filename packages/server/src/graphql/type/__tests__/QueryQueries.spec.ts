import { graphql } from 'graphql';

import { sanitizeTestObject } from '@booksapp/test-utils';

import { schema } from '../../schema';

import {
  getContext,
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  createUser,
  gql,
} from '../../../../test/helpers';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('Query queries', () => {
  it('should get me from query', async () => {
    const user = await createUser();

    const query = gql`
      query Q {
        me {
          id
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });

    const variables = {};

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });
});
