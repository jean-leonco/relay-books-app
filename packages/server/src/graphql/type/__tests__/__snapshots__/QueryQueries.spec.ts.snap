// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query queries should get me from query 1`] = `
Object {
  "me": Object {
    "id": "FROZEN-ID",
  },
}
`;
